### PR TITLE
GDScript: Allow constant expressions in annotations

### LINKED ISSUE #41183: GDScript 2.0: @export_range doesn't allow negative values for the start of a range.
___

***Bugsquad note:** This issue has been confirmed several times already. No need to confirm it further.*

___

**Godot version:**
Godot Engine v4.0.dev.custom_build.a32b427f9

**OS/device including version:**
Windows 10 (64-bit)

**Issue description:**
@export_range does not allow the starting value to be negative.

``` gdscript
extends Node

# Does not allow negative values for start of range.
@export_range(-1, 1, 0.05) var exported_variable: float;

# Zero is allowed.
@export_range(0, 1, 0.05) var working_exported_variable: float;

# Positive is allowed.
@export_range(1, 50, 0.05) var another_working_exported_variable: float;

func _ready():
	pass 
```

Error encountered:
`error(4,28): Expected float as argument 1 of annotation "@export_range").`

**Steps to reproduce:**
Add example script in the issue description to project.

**Minimal reproduction project:**
[TestGDScript2_export_range.zip](https://github.com/godotengine/godot/files/5054826/TestGDScript2_export_range.zip)

### PR DESCRIPTION
**Pull request by dalexeev**
This PR moves annotation resolution from the parser to the analyzer.

For example, the following code now work:

```gdscript
const T = 1

@export_range(-1, 10) var a := 5
@export_range(1 + 1, 10) var b := 5
@export_range(T, 10) var c := 5
```

Closes #41183.

### CODE REVIEW

**Comment by dalexeev** [comment]
Annotations have a feature: strings that are valid identifiers can be written without quotes. This results in the following non-obvious behavior:

```gdscript
const TEXT = "Hello World"

@export_placeholder(TEXT) var s1 = ""
# This is equivalent to:
# @export_placeholder("TEXT") var s1 = ""

# Within an expression, this is correctly recognized as an identifier.
@export_placeholder(TEXT + "") var s2 = ""
```



Options:

1. Leave it as is.
2. Always require quotes.
3. Only allow this in a few annotations (`@warning_ignore`, `@export_node_path`).
4. Check if an identifier exists. If it exists, do not replace it with a string.

**Comment by Garmelon** [comment]
This inconsistent identifier handling seems like it could cause confusion weird bugs in the future. As someone who hasn't used Godot a lot, I wouldn't expect literals to be interpreted as strings.

One example that might come up quite a bit is using a previously defined constant as the bound for some range.
```gdscript
const MAX_SPEED = 4.5

@export_range(0, MAX_SPEED)
var speed
```

Looking at the current documentation, only the examples for [`@export_node_path`](https://docs.godotengine.org/en/latest/classes/class_%40gdscript.html#class-gdscript-annotation-export-node-path) use literals like this. In the other annotation examples, normal string literals are used.

I'm for option 2 (always require quotes).

Option 3 (allow only in some annotations) makes the behavior even more inconsistent (you must remember that this behavior exists *and* you must remember in which annotations it exists) for only a little bit of convenience.

Option 4 (allow if no such identifier exists) would mean that introducing a new constant or variable could change the behavior of an unrelated part of your code just because you happened to name something the same. I can see this happening if somebody uses `@export_enum(DOG, CAT) var companion` and later defines a `const DOG = "Stella", for example`.

**Comment by JoNax97** [comment]
I vote for option 2

**Comment by dalexeev** [comment]
@vnen We need your opinion on https://github.com/godotengine/godot/pull/71634#issuecomment-1396599109.

**Comment by dalexeev** [comment]
Implemented option 2.

File: doc/classes/Dictionary.xml [hunk 1]
```diff
@@ -42,7 +42,7 @@
 		You can access a dictionary's value by referencing its corresponding key. In the above example, [code]points_dict["White"][/code] will return [code]50[/code]. You can also write [code]points_dict.White[/code], which is equivalent. However, you'll have to use the bracket syntax if the key you're accessing the dictionary with isn't a fixed string (such as a number or variable).
 		[codeblocks]
 		[gdscript]
-		@export(String, "White", "Yellow", "Orange") var my_color
+		@export_enum("White", "Yellow", "Orange") var my_color: String
```
Comment (vnen) [inline doc/classes/Dictionary.xml:45]: This is not supported on master yet (see #54828). Though the old version was wrong anyway so this is an improvement.

File: modules/gdscript/gdscript_analyzer.cpp [hunk 2]
```diff
@@ -1290,7 +1295,53 @@ void GDScriptAnalyzer::resolve_node(GDScriptParser::Node *p_node, bool p_is_root
...
+				return;
+			}
+
+			Variant converted_to;
+			const Variant *converted_from = &value;
+			Callable::CallError call_error;
+			Variant::construct(argument_info.type, converted_to, &converted_from, 1, call_error);
+
+			// This should not happen.
+			ERR_FAIL_COND_MSG(call_error.error != Callable::CallError::CALL_OK, vformat(R"(Cannot convert argument %d of annotation "%s" from "%s" to "%s".)", i + 1, p_annotation->name, Variant::get_type_name(value.get_type()), Variant::get_type_name(argument_info.type)));
```
Comment (vnen) [inline modules/gdscript/gdscript_analyzer.cpp:1338]: This should also (or only) do a `push_error()` so the script isn't successfully compiled if this fails, even if it isn't supposed to (or not checked at all if it's impossible).

**Review by vnen — APPROVED** [review]
LGTM. The comment I made is not blocking.

**Comment by akien-mga** [comment]
Thanks!

**Comment by vaartis** [comment]
Somehow this PR breaks `@export var: CustomType`, since now it shows "Parse Error: Identifier "(custom type name)" not declared in the current scope.". Before this PR this only happened on first project import, but now it's always happening.

**Comment by dalexeev** [comment]
Perhaps the reason is the change in `GDScriptLanguage::get_global_class_name()`. Check `.godot/global_script_class_cache.cfg`. Probably the `native_type` of the class is not properly resolved in case of deep inheritance (bug in the analyzer).

**Comment by vaartis** [comment]
[Minimal example of the issue](https://github.com/godotengine/godot/files/10528184/Test.zip)

**Pull request merged**